File: misc/samples/js/src/strands/runner.ts
// IMPORTANT: Telemetry must be set up BEFORE importing samples (which load AWS SDK)
// This ensures AWS SDK instrumentation captures all calls
import {
  DEFAULT_MODEL,
  MODEL_ALIASES,
  REASONING_MODELS,
  DEFAULT_THINKING_BUDGET,
} from '../shared/config.js';
import { setupTelemetry, shutdownTelemetry, Frameworks } from '../shared/telemetry.js';
import { createTraceAttributes } from '../shared/trace.js';

export interface SampleOptions {
  enableThinking?: boolean;
  thinkingBudget?: number;
}

type Sample = {
  run: (modelId: string, options?: SampleOptions) => Promise<void>;
};

// Samples loaded dynamically after telemetry setup
let SAMPLES: Record<string, Sample> = {};

// Convert kebab-case to camelCase for sample lookup
const toCamelCase = (s: string) => s.replace(/-([a-z])/g, (_, c) => c.toUpperCase());

// Convert camelCase to kebab-case for display
const toKebabCase = (s: string) =>
  s
    .replace(/([A-Z])/g, '-$1')
    .toLowerCase()
    .replace(/^-/, '');

async function loadSamples(): Promise<void> {
  const samples = await import('./samples/index.js');
  SAMPLES = samples as unknown as Record<string, Sample>;
}

function printHelp() {
  console.log('Usage: npm run strands -- <sample> [options]');
  console.log('\nOptions:');
  console.log('  --model=<alias>  Model alias or full model ID (default: bedrock-haiku)');
  console.log('  --sideseat       Use SideSeat SDK for telemetry');
  console.log('  --list           List available samples and model aliases');
  console.log('  --help           Show this help message');
  console.log('\nSamples: Use --list to see available samples');
}

function printList() {
  console.log('Available Samples:');
  console.log('-'.repeat(50));
  for (const name of Object.keys(SAMPLES)) {
    console.log(`  ${toKebabCase(name)}`);
  }
  console.log();

  console.log('Model Aliases:');
  console.log('-'.repeat(50));
  for (const [alias, modelId] of Object.entries(MODEL_ALIASES)) {
    console.log(`  ${alias.padEnd(20)} -> ${modelId}`);
  }
  console.log();
  console.log(`Default: ${DEFAULT_MODEL}`);
}

async function runSample(
  name: string,
  sample: Sample,
  modelArg: string,
  useSideseat: boolean
): Promise<boolean> {
  // Create trace attributes for logging (session ID display)
  const traceAttrs = createTraceAttributes(name);

  console.log(`Running sample: ${toKebabCase(name)}`);
  console.log(`  Model: ${modelArg}`);
  console.log(`  SideSeat telemetry: ${useSideseat}`);
  console.log(`  Session: ${traceAttrs['session.id']}`);

  // Enable extended thinking for reasoning sample
  const enableThinking = name === 'reasoning' && REASONING_MODELS.has(modelArg);
  if (enableThinking) {
    console.log(`  Extended thinking: enabled (budget=${DEFAULT_THINKING_BUDGET} tokens)`);
  }
  console.log();

  const options: SampleOptions = enableThinking
    ? { enableThinking: true, thinkingBudget: DEFAULT_THINKING_BUDGET }
    : {};

  await sample.run(modelArg, options);
  return true;
}

async function main() {
  const args = process.argv.slice(2);

  // Parse flags
  const useSideseat = args.includes('--sideseat');
  const showList = args.includes('--list');
  const showHelp = args.includes('--help') || args.includes('-h');
  const modelArg = args.find((a) => a.startsWith('--model='))?.split('=')[1] ?? DEFAULT_MODEL;
  const rawName = args.find((a) => !a.startsWith('--'));

  if (showHelp) {
    printHelp();
    return;
  }

  // Setup telemetry BEFORE loading samples (which import AWS SDK)
  // This ensures AWS SDK instrumentation captures all Bedrock calls
  setupTelemetry({ useSideseat, framework: Frameworks.Strands });

  // Now load samples (which will import Strands SDK -> AWS SDK)
  await loadSamples();

  if (showList) {
    printList();
    return;
  }

  if (!rawName) {
    printHelp();
    process.exit(1);
  }

  // Normalize sample name: tool-use -> toolUse
  const sampleName = toCamelCase(rawName);

  if (rawName === 'all') {
    const results: Array<{ name: string; ok: boolean; error?: string }> = [];
    for (const [name, sample] of Object.entries(SAMPLES)) {
      console.log(`\n${'='.repeat(60)}\nRunning: ${name}\n${'='.repeat(60)}`);
      try {
        await runSample(name, sample, modelArg, useSideseat);
        results.push({ name, ok: true });
        console.log(`[OK] ${name}`);
      } catch (e) {
        results.push({ name, ok: false, error: String(e) });
        console.error(`[FAILED] ${name}:`, e);
      }
    }
    // Summary
    console.log(`\n${'='.repeat(60)}\nSummary\n${'='.repeat(60)}`);
    const passed = results.filter((r) => r.ok).length;
    const failed = results.length - passed;
    console.log(`Passed: ${passed}/${results.length}, Failed: ${failed}`);
    await shutdownTelemetry();
    if (failed > 0) process.exit(1);
    return;
  }

  const sample = SAMPLES[sampleName];
  if (!sample) {
    console.error(`Unknown sample: ${sampleName}`);
    console.error('Available:', Object.keys(SAMPLES).map(toKebabCase).join(', '));
    process.exit(1);
  }

  await runSample(sampleName, sample, modelArg, useSideseat);

  // Flush pending traces before exit
  await shutdownTelemetry();
}

main().catch(async (e) => {
  console.error('Fatal error:', e);
  await shutdownTelemetry();
  process.exit(1);
});

File: misc/samples/python/src/agent_framework_sample/runner.py
"""Sample runner with model and provider configuration."""

import asyncio

from agent_framework_sample.config import MODEL_ALIASES, REASONING_MODELS, SAMPLES
from agent_framework_sample.telemetry_setup import setup_telemetry
from common.models import DEFAULT_THINKING_BUDGET
from common.runner import create_trace_attributes, run_all_samples_base


def get_client(model_alias: str, enable_thinking: bool = False):
    """Create an Agent Framework client from alias or full model ID.

    Returns an OpenAIChatClient, OpenAIResponsesClient (for reasoning), or
    AnthropicClient depending on the provider.
    """
    if model_alias in MODEL_ALIASES:
        provider, model_id = MODEL_ALIASES[model_alias]
    else:
        if model_alias.startswith("openai-"):
            provider = "openai"
            model_id = model_alias[7:]
        elif model_alias.startswith("anthropic-"):
            provider = "anthropic"
            model_id = model_alias[10:]
        else:
            provider = "openai"
            model_id = model_alias

    thinking_supported = model_alias in REASONING_MODELS
    use_thinking = enable_thinking and thinking_supported

    if provider == "openai":
        # Use OpenAIResponsesClient for reasoning (supports reasoning_effort)
        # Use OpenAIChatClient for standard tasks
        if use_thinking:
            from agent_framework.openai import OpenAIResponsesClient

            print("  Extended thinking: enabled (reasoning_effort=medium)")
            return OpenAIResponsesClient(model_id=model_id)
        else:
            from agent_framework.openai import OpenAIChatClient

            return OpenAIChatClient(model_id=model_id)

    elif provider == "anthropic":
        from agent_framework.anthropic import AnthropicClient

        if use_thinking:
            print(f"  Extended thinking: enabled (budget={DEFAULT_THINKING_BUDGET} tokens)")
        return AnthropicClient(model_id=model_id)

    raise ValueError(f"Unknown provider: {provider}")


def run_sample(name: str, args) -> bool:
    """Run a single sample with the specified configuration."""
    if name not in SAMPLES:
        print(f"Unknown sample: {name}")
        return False

    print(f"Running sample: {name}")
    print(f"  Model: {args.model}")
    print(f"  SideSeat telemetry: {args.sideseat}")
    print()

    setup_telemetry(use_sideseat=args.sideseat)

    enable_thinking = name == "reasoning"
    client = get_client(args.model, enable_thinking=enable_thinking)
    trace_attrs = create_trace_attributes("agent-framework", name)

    import importlib

    module = importlib.import_module(SAMPLES[name])

    extra_kwargs: dict = {}
    if name == "reasoning":
        from agent_framework_sample.config import MODEL_ALIASES

        if args.model in MODEL_ALIASES:
            provider, _ = MODEL_ALIASES[args.model]
        else:
            provider = "openai" if args.model.startswith("openai-") else "anthropic"
        extra_kwargs["provider"] = provider

    asyncio.run(module.run(client, trace_attrs, **extra_kwargs))
    return True


def run_all_samples(args):
    """Run all samples in sequence."""
    run_all_samples_base(SAMPLES, run_sample, args)
